Mouse events for swiping

robrkerr opened the pull request:
I've added mouse events that mimic the corresponding touch events and reuse the corresponding functions. This was something I needed (and also brought up in #254). It appears to be working for me but let me know if you see any issues with the code or if you have suggestions for testing it. 

Closes #254.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -746,6 +756,17 @@ class SwipeableViews extends Component {
       onTouchEnd: this.handleTouchEnd,
+      onMouseDown: (e) => {
+        e.preventDefault();
+        e.persist();
```
I don't think that we need to persist it as it's synchronous.

robrkerr replied:
I was getting a react warning about modifying a shared event otherwise. This stopped the warning from appearing but maybe there is a better way.

oliviertassinari replied:
Oh, that's because of the following line 👍  (`.touches =`)

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -746,6 +756,17 @@ class SwipeableViews extends Component {
       onTouchStart: this.handleTouchStart,
       onTouchEnd: this.handleTouchEnd,
+      onMouseDown: (e) => {
+        e.preventDefault();
```
Why preventing the default? Is it for the text selection? Could you add a comment?

robrkerr replied:
In my case, I had an image in the panel that has a default drag behaviour. This was to prevent that from happening.

oliviertassinari replied:
I think that it should be users responsibility to prevent that behavior.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -746,6 +756,17 @@ class SwipeableViews extends Component {
     const touchEvents = disabled ? {} : {
       onTouchStart: this.handleTouchStart,
       onTouchEnd: this.handleTouchEnd,
+      onMouseDown: (e) => {
```
`event`

robrkerr replied:
Sure

robrkerr commented:
Thanks for the quick response! And good suggestions. So I'll put it behind an option that is off by default. 

And regarding your second point: are you worried that touch events on mobile will be triggering by touch and mouse events (and therefore calling these functions twice)? If that's what you meant, I can look into that and see if there's a nice way to avoid that.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -389,6 +389,14 @@ class SwipeableViews extends Component {
         passive: false,
       },
     );
+    this.mouseMoveListener = addEventListenerEnhanced(this.rootNode, 'mousemove', (e) => {
```
Do we need the event to be `active` (passive: false)? I would expect a regular `onMouseMove` callback property to be enough.

robrkerr commented:
Ok, so I've moved this feature behind a flag 'enableMouseEvents' so that it is disabled by default. However, I'd be inclined to have it enabled by default so that you get the same behaviour on mobile and desktop - don't you think that would be what people would tend to expect?

I've also added separate callbacks for the mouse events and made sure that only the appropriate callback is triggered for the different events. I've called it 'onMouseDrag' instead of 'onMouseMove' because it is only gets triggered while the user is holding down on the mouse button (dragging) - consistent with the 'onTouchMove' behaviour.

robrkerr commented:
I found an issue with the mouse leaving the swipeable region and subsequent 'mouseup' events being missed. This is one solution: end the mouse drag action when this happens. Another would be to try and catch the 'mouseup' events regardless of where the mouse ends up. I think this solution is fine but let me know what you think.

oliviertassinari commented:
@robrkerr The mouse leave approach sounds good to me.

robrkerr commented:
Just realised I should have been using the enhanced event listener - fixed this now.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -255,6 +255,11 @@ class SwipeableViews extends Component {
+     * If `true`, it will enable mouse events.
+     * This will allow the user to perform the relevant swipe actions with a mouse.
+     */
+    enableMouseEvents: PropTypes.bool,
```
Can we add a default value for that property? Just for the documentation purposes.

robrkerr replied:
Done.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -389,6 +406,13 @@ class SwipeableViews extends Component {
         passive: false,
       },
     );
+    this.mouseMoveListener = addEventListenerEnhanced(this.rootNode, 'mousemove', (event) => {
```
If we don't need the listener to be passive, can we use the *idiomatic* way. I mean like `onMouseUp` for instance.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -744,8 +765,38 @@ class SwipeableViews extends Component {
+        if (onTouchEnd) onTouchEnd(event);
+        this.handleTouchEnd(event);
+      },
+      onMouseDown: (event) => {
```
Defining the handler function in the render method is affecting performances. Using a `handleTouchStart` is preferable.

oliviertassinari commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -744,8 +765,38 @@ class SwipeableViews extends Component {
+      },
+      onMouseLeave: (event) => {
+        if (enableMouseEvents && this.started) {
+          if (onMouseUp) onMouseUp(event);
```
Calling `onMouseUp` in a `onMouseLeave` listener? What's the logic/consistency here?

oliviertassinari replied:
We should be calling `onMouseLeave` no matter if `enableMouseEvents` is enabled or not. Otherwise the event is swallowed for our users.

robrkerr commented on packages/react-swipeable-views/src/SwipeableViews.js:
```diff
@@ -732,6 +788,13 @@ class SwipeableViews extends Component {
+      onMouseLeave, // eslint-disable-line no-unused-vars
+      onMouseMove, // eslint-disable-line no-unused-vars
+      onMouseUp, // eslint-disable-line no-unused-vars
+      enableMouseEvents, // eslint-disable-line no-unused-vars
```
I guess these 'onTouch/Mouse' handlers don't need to be remove here because they will simply get overwritten right?

oliviertassinari replied:
I agree, it's even an issue when `disabled` is true.

robrkerr commented:
@oliviertassinari: Thanks again for your feedback! I think I've fixed up everything you were concerned about - please let me know if I haven't. 

I've also renamed the original 'handleTouchStart', 'handleTouchMove', and 'handleTouchEnd' functions (which can now be target by either touch or mouse events) to 'handleSwipeStart', 'handleSwipe', and 'handleSwipeEnd'. What do you think of these names? Maybe 'handleSwipe' would be better as 'handleSwipeMove'?

Is there anything else you think needs fixing / changing?

oliviertassinari commented:
@robrkerr Thanks. I will have a deep look into it tomorrow.

oliviertassinari commented:
@robrkerr Thanks for the effort. I'm continuing it with #266.
